### DISCUSSION TITLE
what is `cue import openapi FILE` supposed to do?

### DISCUSSION THREAD

**Discussion by BrianHicks** [cue-lang/cue#1532] [Q&A]
I'm trying to import an OpenAPI file to CUE and finding that it doesn't do what I expected: I thought importing it would import the things in the `definitions` field as CUE schemas, but it looks like it just imports the OpenAPI file as JSON (including things like setting the `description` field, etc.

Am I way off base here? I thought this would be a conversion, not a straight import—in fact, `cue import json FILE` seems to produce identical results to `cue import openapi FILE`.

To be completely open about this: I couldn't figure it out and have written something that does what I intended (that is, convert OpenAPI defintions to CUE schemas) over the past couple of days. I would like to open-source it, but I don't want to create a big conflict between my thing and the official language tooling!

**Comment by xinau** [comment]
Hi @BrianHicks, would it be possible for you to specify a bit more about the exact commands and filenames you've used for importing the OpenAPI schema. 

Using a random OpenApi schema from [DZone](https://dzone.com/articles/a-sample-openapi-30-file-to-get-started) and saving it as `test.openapi.yml`, I'm able to import (convert) it as a CUE schema.

```console
$ cue import -o - test.openapi.yaml
// Swagger Petstore

info: {
        version: *"1.0.0" | string
        title:   *"Swagger Petstore" | string
        license: name: "MIT"
}

#Pet: null | bool | number | string | [...] | {
        id:   int
        name: string
        tag?: string
        ...
}
#Pets: [...#Pet]
#Error: null | bool | number | string | [...] | {
        code:    int
        message: string
        ...
}
```
```console
$ cue import openapi
$ cat test.openapi.cue
// Swagger Petstore
...
```

In case your filenames don't have a extension from which an OpenAPI schema can be determined it's possible to specify the filetype.

```console
$ cue import -o - openapi: test.yaml
// Swagger Petstore
...
```

There's also a [test case](https://github.com/cue-lang/cue/blob/7cdcc70bee6a31defc94f8a719d75bbbdcbc5680/cmd/cue/cmd/testdata/script/import_auto.txt#L1) for the cue import command that uses OpenApi as input. I hope these information's already help a bit.

**Reply by BrianHicks** [reply]
It's the Kubernetes OpenAPI spec and a few others (Argo Workflows, CD, and Rollouts.) I know I could hypothetically do `cue get go` on these, but the dependency management with all those dependencies as Go was jut not working for me.

Before, I've run `cue import openapi swagger.json` on https://github.com/kubernetes/kubernetes/raw/master/api/openapi-spec/swagger.json and gotten something that looks like this:

```cue
definitions: {
	"io.k8s.api.admissionregistration.v1.MutatingWebhook": {
		description: "MutatingWebhook describes an admission webhook and the resources and operations it applies to."
	}
	...
}
```

However, after reading your comment I added a colon after the format… and I see now in `cue help filetypes` that the colon is required. (Side note: it's weird that `cue import openapi foo.json` succeeds when the `openapi` file doesn't exist? Is that what that command is actually saying? Why doesn't it show an error?)

Anyway, one I add that colon it still gives a failure:

```
value "components" not found
```

I see that the K8s spec uses openapi 2 but the example you loaded successfully is openapi 3. Do you know the earliest version CUE supports? I don't see it in the help.

**Reply by xinau** [reply]
I found the cli behavior of import myself a bit non intuitive and think it could be improved up on.

As I'm not that fond with OpenApi I'm unable to answer your last question, but I would guess it doesn't support version 2. But maybe someone with more knowledge on this topic can answer it.

**Comment by jornh** [comment]
Regarding what OpenAPI version is supported the Go API doc hints at 3.0.0 here:
https://pkg.go.dev/cuelang.org/go@v0.4.2/encoding/openapi#pkg-overview

A related issue regarding import/export doc improvements: https://github.com/cue-lang/cue/issues/1012
